src/NumericalAlgorithms/Spectral/Spectral.hpp

```diff
@@ -0,0 +1,119 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/*!
+ * \file
+ * Declares functions to retrieve spectral quantities associated with a
+ * particular choice of basis functions and quadrature method.
+ */
+
+#pragma once
+
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+
+class Matrix;
+
+/*!
+ * \ingroup SpectralGroup
+ * \brief Basis functions, derivative matrices, etc. for spectral-type methods
+ */
+namespace Spectral {
+
+/*!
+ * \ingroup SpectralGroup
+ * \brief The choice of collocation points and weights to compute an inner
+ * product (integral) between basis functions, such that they are orthogonal
+ *
+ * \detail For each member of `Spectral::Basis` and `Spectral::Quadrature`,
+ * the class `Spectral::detail::CollocationPointsAndWeights` should implement
+ * a template-specialized constructor.
+ */
+enum class Quadrature { Gauss, GaussLobatto };
+
+/*!
+ * \brief Collocation points
+ * \param number_of_pts number of collocation points
+ */
+template <typename BasisType>
+const DataVector& collocation_points(size_t number_of_pts) noexcept;
+
+/*!
+ * \brief Quadrature weights
+ * \param number_of_pts number of collocation points
+ */
+template <typename BasisType>
+const DataVector& quadrature_weights(size_t number_of_pts) noexcept;
+
+/*!
+ * \brief Matrix used to compute the first derivative of a function.
+ * \param number_of_pts number of collocation points
+ */
+template <typename BasisType>
+const Matrix& differentiation_matrix(size_t number_of_pts) noexcept;
+
+/*!
+ * \brief Matrix used to compute spectral coefficients of a function.
+ * \param number_of_pts number of collocation points
+ */
+template <typename BasisType>
+const Matrix& grid_points_to_spectral_matrix(size_t number_of_pts) noexcept;
+
+/*!
+ * \brief Matrix used to compute function at the collocation points from its
+ * spectral coefficients.
+ * \param number_of_pts number of collocation points
+ */
+template <typename BasisType>
+const Matrix& spectral_to_grid_points_matrix(size_t number_of_pts) noexcept;
+
+/*!
+ * \brief Matrix used to linearize a function.
+ *
+ * \details Filters out all except the lowest two modes.
+ * \param number_of_pts number of collocation points
+ */
+template <typename BasisType>
+const Matrix& linear_filter_matrix(size_t number_of_pts) noexcept;
+
+/*!
+ * \brief Matrix to interpolate to a set of target points
+ *
+ * \warning It is expected but not checked that the target points are inside
+ * the domain.
+ * \param number_of_pts number of collocation points
+ * \param target_points the points to interpolate to
+ */
+template <typename BasisType, typename T>
+Matrix interpolation_matrix(size_t number_of_pts,
+                            const T& target_points) noexcept;
+
+/*!
+ * \brief Maximum number of allowed collocation points
+ */
+static const size_t maximum_number_of_pts = 12;
+
+namespace detail {
+template <typename BasisType>
+class CollocationPointsAndWeights {
```

Comment:
can be moved to a cpp file

Reply:
I'm not quite sure but I think not, since we want to specialise it in Basis-specific cpp files such as `Legendre.cpp` (to provide a specialised constructor). I could move it to `Spectral.tpp` I guess, which is included in each of these Basis-specific files anyway.